### PR TITLE
[EVM] Add memory, context, system and log instructions.

### PR DESCRIPTION
**Pull request by PavelKopyl**
This includes:
 - memory (heap) instructions
 - storage instructions/intrinsics
 - retrieval values from the context instructions/intrinsics
 - call* instructions/intrinsics
 - logging instructions/intrinsics

### CODE REVIEW

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

File: llvm/include/llvm/IR/IntrinsicsEVM.td [hunk 1]
```diff
@@ -17,4 +17,152 @@ def int_evm_mulmod
...
+  : Intrinsic<[], [LLVMPointerType<llvm_i256_ty>, llvm_i256_ty, llvm_i256_ty],
+              [IntrWriteMem]>;
+
+def int_evm_codesize : Intrinsic<[llvm_i256_ty], [], [IntrNoMem]>;
+
+def int_evm_codecopy
+  : Intrinsic<[], [LLVMPointerType<llvm_i256_ty>, llvm_i256_ty, llvm_i256_ty],
+              [IntrWriteMem]>;
+
+def int_evm_gasprice : Intrinsic<[llvm_i256_ty], [], [IntrNoMem]>;
```
Comment (akiramenai) [inline llvm/include/llvm/IR/IntrinsicsEVM.td:70]: I don't know if gas price can change between calls if it called twice in the same function.

**Review by akiramenai — APPROVED** [review]

File: llvm/include/llvm/IR/IntrinsicsEVM.td (hunk 1 above)
Comment (PavelKopyl) [reply llvm/include/llvm/IR/IntrinsicsEVM.td:70]: As I understand, gasprice should not be changed between in the same function, as it is determined by the protocol.
I'm not 100% sure, so I'll add TODO in the comments to check if this is true?

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

**Pull request merged**